keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,146 @@
+
+As the discovery document now exposes the [storage version hash][], storage version
+migration should start automatically when the hash changes.
+
+[storage version hash]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md
+
+## API design
+
+We introduce the `StorageState` API, as a CRD defined in the `migration.k8s.io`
+group.
+
+```golang
+// StorageState is the state of the storage, for a specific resource.
+type StorageState struct {
+  metav1.TypeMeta
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StoageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of versions that the persisted objects might be encoded in.
+  // If empty, it means the stored versions are unknown.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we make it part of the status because
+* Its value can be determined by inspecting the `StorageVersionMigration` API.
+  It just takes longer, i.e., until a migration has completed. 
+* Leaving it in the status is consistent with the [CRD.status.StoredVersions][]
+  API.
+
+[CRD.status.StoredVersions]:https://github.com/kubernetes/kubernetes/blob/697c2316faaabae8ef8371032b60be65d7795e68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305
+  
+## Storage migration triggering controller
+
+We will add a controller that monitors the discovery documents and the
+storageStates to trigger storage version migration.
+
+The controller fetches the discovery document periodically. For each resource,
+it compares the storageVersion in the discovery document with that recorded in
+the storageState. If they differ, or if the storageState does not exist, the
+controller
+  * stops any ongoing migration of this resource by deleting existing
+    storageVersionMigrations,
+  * creates a new storageVersionMigration for this resource,
+  * in a single operation, updates (or creates) the
+    storageState.status.StorageVersionHash to the value shown in the discovery
+    document, and adds the hash to the storageState.status.storedVersionHashes.
+
+The refresh rate of the discovery document needs to satisfy two conflicting
+goals: *i)* not causing too much traffic to apiservers, and *ii)* not missing
+storage version configuration changes. 10 minutes is a good middle ground.
```

Comment:
so if it missed the change, it could wait there forever, without triggering the migration?

Reply:
"missing storage version configuration changes" means the storage version of a resource has changed twice in a refresh period. If so, migration will never be triggered and it's not safe. Reworded.

I had highlighted this in the "Implications to cluster operators" that admins should upgrade/downgrade the cluster twice in 10 mins.